src/Microsoft.AspNetCore.Analyzer.Testing/DiagnosticVerifier.cs

```diff
@@ -0,0 +1,189 @@
+
+                foreach (var diag in diags)
+                {
+                    if (diag.Location == Location.None || diag.Location.IsInMetadata)
+                    {
+                        diagnostics.Add(diag);
+                    }
+                    else
+                    {
+                        foreach (var document in documents)
+                        {
+                            var tree = await document.GetSyntaxTreeAsync();
+                            if (tree == diag.Location.SourceTree)
+                            {
+                                diagnostics.Add(diag);
+                            }
+                        }
+                    }
+                }
+            }
+
+            return diagnostics.OrderBy(d => d.Location.SourceSpan.Start).ToArray();
+        }
+
+        /// <summary>
+        /// Given an array of strings as sources and a language, turn them into a project and return the documents and spans of it.
+        /// </summary>
+        /// <param name="sources">Classes in the form of strings</param>
+        /// <returns>An array of Documents produced from the sources.</returns>
+        private Document[] GetDocuments(string[] sources)
+        {
+            var project = CreateProject(sources);
+            var documents = project.Documents.ToArray();
+
+            if (sources.Length != documents.Length)
+            {
+                throw new InvalidOperationException("Amount of sources did not match amount of Documents created");
+            }
+
+            return documents;
+        }
+
+        /// <summary>
+        /// Create a project using the inputted strings as sources.
+        /// </summary>
+        /// <param name="sources">Classes in the form of strings</param>
+        /// <returns>A Project created out of the Documents created from the source strings</returns>
+        private Project CreateProject(string[] sources)
+        {
+            string fileNamePrefix = DefaultFilePathPrefix;
+
+            var projectId = ProjectId.CreateNewId(debugName: TestProjectName);
+
+            var solution = new AdhocWorkspace()
+                .CurrentSolution
+                .AddProject(projectId, TestProjectName, TestProjectName, LanguageNames.CSharp);
+
+            foreach (var defaultCompileLibrary in DependencyContext.Load(this.GetType().Assembly).CompileLibraries)
+            {
+                foreach (var resolveReferencePath in defaultCompileLibrary.ResolveReferencePaths(new AppLocalResolver()))
+                {
+                    solution = solution.AddMetadataReference(projectId, MetadataReference.CreateFromFile(resolveReferencePath));
+                }
+            }
+
+            int count = 0;
+            foreach (var source in sources)
+            {
+                var newFileName = fileNamePrefix + count;
+                var documentId = DocumentId.CreateNewId(projectId, debugName: newFileName);
```

Comment:
Do we want to special case `sources.Length == 0` so that `newFileName == fileNamePrefix`? Just easier to work with it